File: algospot/questack/BRACKETS2/main.cpp
#include <iostream>
#include <stack>
#include <string.h>

using namespace std;

bool isOpener(char c);
bool isProperCloser(char open, char close);

int main()
{
    int testCase;
    cin >> testCase;
    while(testCase--) {
        stack<char> bracket;
        bool result = true;
        string b;
        cin >> b;
        int len = b.length();
        
         for(int i=0; i<len; i++) {
             // open 괄호면 push한다
            if(isOpener(b[i])) {
                bracket.push(b[i]);
            } else {
                // !bracket.empty() : push된 open 괄호가 없으면 안된다
                // isProperCloser : top에 있는 open 괄호와 맞는 close 괄호인지 확인한다
                if(!bracket.empty() && isProperCloser(bracket.top(), b[i])) {
                    bracket.pop();
                } else {
                    result = false;
                    break;
                }
            }
        }
        
        // push된 open 괄호를 전부 써야 맞다
        if(!bracket.empty()) {
            result = false;
        }
        if(result) cout << "YES" << endl;
        else cout << "NO" << endl;
    }
    return 0;
}

bool isOpener(char c)
{
    if(c == '(' || c == '{' || c == '[') {
        return true;
    }
    return false;
}

bool isProperCloser(char open, char close)
{
    switch(open) {
        case '(' :
            if(close == ')') return true;
            else return false;
        case '{' :
            if(close == '}') return true;
            else return false;
        case '[' :
            if(close == ']') return true;
            else return false;
        default :
            return false;
    }
}

File: algospot/questack/BRACKETS2/main.py
class Stack:
    def __init__(self):
        self.items = []

    def isEmpty(self):
        return self.items == []

    def push(self, item):
        self.items.append(item)

    def pop(self):
        return self.items.pop()

    def top(self):
        return self.items[len(self.items)-1]

def isOpen(c) :
    if c == '(' or c == '[' or c == '{' :
        return True
    return False

def isMatched(opened, closed) :
    if opened == '(' and closed == ')' : return True
    if opened == '{' and closed == '}' : return True
    if opened == '[' and closed == ']' : return True
    return False

if __name__ == '__main__':
    testCase = int(input())
    for i in range(testCase):
        str = input()
        st = Stack()
        result = True
        for c in str:
            if isOpen(c) :
                st.push(c)
            else :
                if st.isEmpty() :
                    result = False
                    break;
                opened = st.top()
                st.pop()
                if isMatched(opened, c) == False :
                    result = False
                    break;
        if st.isEmpty() == False : result = False
        if result : print("YES")
        else : print("NO")
